usr/src/man/man1/pkgsrc-wrapper.1
.\"
.\" This file and its contents are supplied under the terms of the
.\" Common Development and Distribution License ("CDDL"), version 1.0.
.\" You may only use this file in accordance with the terms of version
.\" 1.0 of the CDDL.
.\"
.\" A full copy of the text of the CDDL should have accompanied this
.\" source.  A copy of the CDDL is also available via the Internet at
.\" http://www.illumos.org/license/CDDL.
.\"
.\"
.\" Copyright 2023 MNX Cloud, Inc.
.\"
.
.TH PKGSRC\-WRAPPER 1 "April 19, 2023"
.
.SH "NAME"
\fBpkgsrc-wrapper\fR \-\- Platform commands supplied by pkgsrc
.
.SH "SYNOPSIS"
.
.nf
/usr/bin/pkgsrc\-wrapper
.
.fi
.
.P
This command should not be run directly\. This command should only be called
where \fBarg[0]\fR is the name of another command to be \fBexec(2)\fR\'d\.
.
.SH "DESCRIPTION"
This is a generic wrapper for supplying lesser\-used commands via pkgsrc rather
than directly in the platform image\. This allows preserving space in the
ram disk while also extending the platform image with additional commands\. The
primary target for this is shells and interpreters\.
.
.P
This works both with base/minimal zone images as well as pkgsrc\-tools installed
in the global zone\. If both are installed, \fB/opt/tools\fR will be preferred over \fB/opt/local\fR\|\.
.
.P
Note: This does not allow operators to arbitrarily extend the SmartOS platform
image because that would require writing to \fB/usr/bin\fR, which is read\-only\.
Rather, SmartOS itself provides hooks for specific commands that can be
satisfied using pkgsrc\.
.
.P
The following platform commands can be satisfied using pkgsrc:
.
.IP "\(bu" 4
csh (by installing tcsh)
.
.IP "\(bu" 4
tcsh
.
.IP "\(bu" 4
zsh
.
.P
Thus, if \fBtcsh\fR is installed via pkgsrc, then you can call \fB/usr/bin/csh\fR,
which would then invoke \fB/opt/{tools,local}/bin/tcsh\fR\|\. This will allow, for
example, \fB/usr/bin/csh\fR to be used as a script interpreter or as an LDAP user\'s
shell in either the global zone or non\-global zones without any additional
special handling\.
.
.SH "SEE ALSO"
.
.nf
  exec(2), shells(5), csh(1), tcsh(1), zsh(1), pkgin(1)
.
.fi
.
.SH "NOTES"
The SmartOS platform image is currently mosty composed of 32\-bit binaries\.
Packages installed via pkgsrc will be 64\-bit if using pkgsrc\-tools (in the
global zone) or a base\-64/minimal\-64 zone image\.


usr/src/cmd/pkgsrc-wrapper/pkgsrc-wrapper.c
/*
 * This file and its contents are supplied under the terms of the
 * Common Development and Distribution License ("CDDL"), version 1.0.
 * You may only use this file in accordance with the terms of version
 * 1.0 of the CDDL.
 *
 * A full copy of the text of the CDDL should have accompanied this
 * source.  A copy of the CDDL is also available via the Internet at
 * http://www.illumos.org/license/CDDL.
 */

/*
 * Copyright 2023 MNX Cloud, Inc.
 */

/*
 * Use a C program instead of a shell script to exec() something in
 * the pkgsrc paths.
 */

#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <assert.h>
#include <errno.h>

/* Remap table for when we have commands with than one name (e.g. tcsh/csh). */
static struct {
	const char *bin_name;
	const char *bin_needs;
} remap_table[] = {
	{ "csh", "tcsh" },
	/* login(1) invokes shells as -$SHELL. Cover that here. */
	{ "-csh", "tcsh" },
	{ "-tcsh", "tcsh" },
	{ "-zsh", "zsh" },
	{ NULL, NULL}
};

static const char *
remap_bin(const char *base_bin)
{
	uint_t index;

	for (index = 0; remap_table[index].bin_name != NULL; index++) {
		/* Can use strcmp() because remap_table entries are bounded. */
		if (strcmp(remap_table[index].bin_name, base_bin) == 0)
			return (remap_table[index].bin_needs);

	}
	return (base_bin);
}


/* List these pkgsrc paths in desired search order. */
static const char *pkgsrc_paths[] = {
	"/opt/local/bin",
	"/opt/local/sbin",
	"/opt/tools/bin",
	"/opt/tools/sbin",
	NULL
};

/*
 * This returns an allocated string.  Since we're exec()-ing or exit()-ing we
 * really don't need to worry about leaks, however.
 */
static char *
generate_pkgsrc_path(const char *desired_bin)
{
	const char *path = NULL;
	char *binpath;
	uint_t index;

	for (index = 0; pkgsrc_paths[index] != NULL; index++) {
		int dirfd;
		int cmdfd;

		/*
		 * NOTE: Any of dirfd or cmdfd that has an actual file
		 * descriptor from open(2) MUST be closed before this loop
		 * either exits or iterates.
		 */

		dirfd = open(pkgsrc_paths[index], O_RDONLY);
		if (dirfd == -1) {
			/* XXX KEBE ASKS, print diagnostics? */
			continue; /* Try next one */
		}
		cmdfd = openat(dirfd, desired_bin, O_RDONLY);
		(void) close(dirfd); /* XXX KEBE WARNS, errno reset... */
		if (cmdfd == -1) {
			/*
			 * XXX KEBE ASKS, print diagnostics?
			 * If so, move close(dirfd) here after printing, or
			 * save off errno for openat() above.
			 */
			continue; /* Try next one */
		}
		(void) close(cmdfd);
		path = pkgsrc_paths[index];
		break;	/* We're done. */
	}

	if (path == NULL)
		return (NULL); /* If we reach here, there's no such binary. */

	/*
	 * Okay, so we can now combine path and desired_bin into a string
	 * suitable for an exec()-family call.
	 */

	if (asprintf(&binpath, "%s/%s", path, desired_bin) == -1) {
		/* XXX KEBE ASKS, print diagnostics? */
		assert(binpath == NULL); /* man pages says this is true. */
	}

	return (binpath);
}

/* Don't let us run ourselves. */
static char *myname = "pkgsrc-wrapper";

int
main(int argc, char *argv[], char *envp[])
{
	const char *base_bin;
	const char *remapped_bin;
	char *pkgsrc_path;

	/*
	 * 1. Get the binary name in argv[0], remapping if need be.
	 * (e.g. csh -> tcsh, or -($SHELL) invocations.)
	 */
	base_bin = strrchr(argv[0], '/');
	if (base_bin == NULL)
		base_bin = argv[0];
	else
		base_bin++; /* Move past matched '/' */

	/*
	 * 2. Make sure we just aren't running ourselves!
	 * (Can use strcmp() because "myname" is bounded.)
	 */
	if (strcmp(myname, base_bin) == 0) {
		/* XXX KEBE ASKS L10N? */
		(void) fprintf(stderr, "%s should not be run by itself.\n",
		    myname);
		exit(3);
	}

	remapped_bin = remap_bin(base_bin);

	/* 3. See if the binary name is in one of the pkgsrc paths. */
	pkgsrc_path = generate_pkgsrc_path(remapped_bin);
	if (pkgsrc_path == NULL) {
		/* XXX KEBE ASKS L10N? */
		/* XXX KEBE ALSO ASKS -- better message? */
		(void) fprintf(stderr, "Please install %s from pkgsrc.\n",
		    remapped_bin);
		exit(1);
	}

	/* 4. If so, launch it with our exact arg[vc] (assume env goes thru?) */
	if (execve(pkgsrc_path, argv, envp) == -1) {
		/* XXX KEBE ASKS L10N? */
		/* XXX KEBE ALSO ASKS -- better message? */
		(void) fprintf(stderr, "Cannot execute %s: %s.\n",
		    pkgsrc_path, strerror(errno));
		exit(2);
	}

	/* We shouldn't return here... */
	/* XXX KEBE ALSO ASKS -- better message? */
	(void) fprintf(stderr, "execve bug with %s.\n", pkgsrc_path);
	return (2);
}
